src/Bicep.Core/TypeSystem/DeployTimeConstantVisitor.cs

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Semantics;
+using Bicep.Core.Syntax;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Bicep.Core.TypeSystem
+{
+    /// <summary>
+    /// Visitor used to collect errors caused by property assignments to run-time values when that is not allowed
+    /// </summary>
+    public sealed class DeployTimeConstantVisitor : SyntaxVisitor
+    {
+
+        // These constants currently resolve to the same name.
+        private readonly HashSet<string> DeployTimeConstantKeys = new HashSet<string>()
+        { 
+            LanguageConstants.ResourceNamePropertyName,
+            LanguageConstants.ModuleNamePropertyName
+        }; 
+
+        private readonly SemanticModel model;
+        private readonly IDiagnosticWriter diagnosticWriter;
+
+        // used for logging the property when we detect a run time property
+        private string currentSymbol;
+
+        // Since we collect errors top down, we will collect and overwrite this variable and emit it in the end
+        private SyntaxBase? errorSyntax;
+
+        private DeployTimeConstantVisitor(SemanticModel model, IDiagnosticWriter diagnosticWriter)
+        {
+            this.model = model;
+            this.diagnosticWriter = diagnosticWriter;
+            this.currentSymbol = "";
```

Comment:
IMO better to make this field nullable and not set it in the constructor. That way the type system will ensure it's used correctly.